android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboard.java

```diff
@@ -185,6 +185,12 @@ protected boolean updateSelectionRange() {
       int selStart = icText.selectionStart;
       int selEnd = icText.selectionEnd;
 
+      if (selStart < 0 || selEnd < 0) {
+        // There is no selection or cursor
+        // Reference https://developer.android.com/reference/android/text/Selection#getSelectionEnd(java.lang.CharSequence)
+        return false;
+      }
+
       int selMin = selStart, selMax = selEnd;
```

Comment:
Personally, I'd like to eliminate the `selStart` and `selEnd` variables. They just add potential for error.

```suggestion
      int selMin = icText.selectionStart, selMax = icText.selectionEnd;

      if (selMin < 0 || selMax < 0) {
        // There is no selection or cursor
        // Reference https://developer.android.com/reference/android/text/Selection#getSelectionEnd(java.lang.CharSequence)
        return false;
      }

      if(selMin > selMax) {
        selMin = icText.selectionEnd;
        selMax = icText.selectionStart;
      }
```

I can't do this suggestion across the whole range of this.

Reply:
Similarly, we should eliminate the `result` variable, and short-circuit the input conditions:

```java
    InputConnection ic = KMManager.getInputConnection(this.keyboardType);
    if (ic == null) {
      // TODO: please explain
      return true;
    }

    ExtractedText icText = ic.getExtractedText(new ExtractedTextRequest(), 0);
    if (icText == null) {
      // TODO: please explain
      return false;
    }

    // ... the rest of the function goes here ...

    return true;
```

Why do we return true when ic == null?

Reply:
Refactoring addressed in #11389. Also fixes to return false if ic == null.